### PR TITLE
backport of changes in #17057 that don't require new worlds

### PR DESCRIPTION
**Pull request by vtjnash**
Fixes some method definition orderings, and other `@generated` functions, that will be broken by PR #17057

### CODE REVIEW

File: base/inference.jl [hunk 1]
```diff
@@ -3923,10 +3931,29 @@ end
...
+        push!(fs, x[3])
+    end
+    for i = 1:length(t_ifunc)
+        if isassigned(t_ifunc, i)
+            x = t_ifunc[i]
+            push!(fs, x[3])
+        end
+    end
+    for f in fs
+        for m in _methods_by_ftype(Tuple{typeof(f), Vararg{Any}}, 10)
```
Comment (tkelman) [inline base/inference.jl]: should this 10 be given a name?

File: base/irrationals.jl [hunk 2]
```diff
@@ -57,27 +57,44 @@ end
...
-    rx < bx ? :($rx < y) : :($rx <= y)
+@pure function rationalize{T<:Integer}(::Type{T}, x::Irrational; tol::Real=0)
+    return rationalize(T, big(x), tol=tol)
 end
-@generated function <{T}(x::Rational{T}, y::Irrational)
-    by = big(y())
-    by < 0 && T <: Unsigned && return false
-    ry = rationalize(T,by,tol=0)
-    ry < by ? :(x <= $ry) : :(x < $ry)
+@pure function rationalize{T<:Integer}(::Type{T}, x::Irrational)
```
Comment (tkelman) [inline base/irrationals.jl]: this causes a method overwrite warning, why is it needed?

File: doc/manual/metaprogramming.rst [hunk 3]
```diff
@@ -958,31 +967,35 @@ used?
...
-*never* write a generated function with side effects - when, and how often,
-the side effects occur is undefined. (This is true for macros too - and just
-like for macros, the use of :func:`eval` in a generated function is a sign that
+Note that there is no printout of :obj:`Int64`. We can see that the body
+of the generated function was only executed once here,
+for the specific set of argument types, and the result was cached.
+After that, for this example, the expression returned from the generated
+function on the first invocation was re-used as the method body.
+The actual caching behavior is an implementation-defined performance optimization,
+so it is invalid to depend to closely on this behavior, however.
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: "too closely"

don't end a sentence with ", however." - just delete that

File: doc/manual/metaprogramming.rst [hunk 4]
```diff
@@ -958,31 +967,35 @@ used?
...
 
 The example generated function ``foo`` above did not do anything a normal
-function ``foo(x)=x*x`` could not do, except printing the type on the
-first invocation (and incurring a higher compile-time cost). However, the
-power of a generated function lies in its ability to compute different quoted
-expression depending on the types passed to it:
+function ``foo(x) = x * x`` could not do (except printing the type on the
+first invocation, and incurring higher overhead).
+However, the power of a generated function lies in its ability to compute
+different quoted expression depending on the types passed to it:
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: different quoted expressions

File: base/inference.jl (hunk 1 above)
Comment (vtjnash) [inline base/inference.jl]: it's just a random round-ish number

File: base/irrationals.jl (hunk 2 above)
Comment (vtjnash) [inline base/irrationals.jl]: for the same reason the other method overwrite warning happens in searchsorted

File: doc/manual/metaprogramming.rst [hunk 5]
```diff
@@ -958,31 +967,35 @@ used?
...
+and how often, the side effects occur is undefined.
+(This is true for macros too - and just like for macros,
+the use of :func:`eval` in a generated function is a sign that
 you're doing something the wrong way.)
+However, unlike macros, the runtime system cannot correctly handle a call to
+:func:`eval`, so it is disallowed.
 
 The example generated function ``foo`` above did not do anything a normal
-function ``foo(x)=x*x`` could not do, except printing the type on the
-first invocation (and incurring a higher compile-time cost). However, the
```
Comment (vtjnash) [inline doc/manual/metaprogramming.rst]: no, it's adds overhead in many places

File: base/irrationals.jl (hunk 2 above)
Comment (tkelman) [inline base/irrationals.jl]: any horizon for not needing that?

Comment (JeffBezanson) [inline base/irrationals.jl]: The one in this case might not be needed --- calling a kwarg function with no keyword arguments is fast. The problem in searchsorted was that computing the keyword arg default values required dynamic dispatch.

Comment (vtjnash) [inline base/irrationals.jl]: True, the problem in this case is actually just that lowering won't apply `@pure` to the kwargs functions, and that's quite fixable.

Comment (tkelman) [inline base/irrationals.jl]: add a TODO comment saying to fix pure for kwargs so the no-kwargs method can be deleted then?

Comment (vtjnash) [inline base/irrationals.jl]: original issue should be corrected by 3f2213f4e74ed3070045da139f659694b7c071a8 now

File: doc/manual/metaprogramming.rst [hunk 6]
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
...
+
 It's easiest to illustrate this with an example. We can declare a generated
 function ``foo`` as
 
 .. doctest::
 
     julia> @generated function foo(x)
-               println(x)
-               return :(x*x)
+               Core.println(x)
```
Comment (mauro3) [inline doc/manual/metaprogramming.rst]: But this is not pure.  This is mentioned way later, but coming after the new point "4." this is just odd.  Maybe state that doing `println` is ok?

File: doc/manual/metaprogramming.rst [hunk 7]
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
...
 
 2. In the body of the generated function you only have access to the
-   *types* of the arguments, not their values.
+   *types* of the arguments -- not their values -- and any function that
+   was defined *before* the definition of generated function.
 
 3. Instead of calculating something or performing some action, you return
    a *quoted expression* which, when evaluated, does what you want.
 
+4. The function must not have any side-effects or examine any non-constant
```
Comment (mauro3) [inline doc/manual/metaprogramming.rst]: To go with the later "they"s, start with "4. Generated functions must not..."

File: doc/manual/metaprogramming.rst [hunk 8]
```diff
@@ -893,37 +893,46 @@ the arguments are known, but the function is not yet compiled.
...
 ordinary functions:
 
 1. You annotate the function declaration with the ``@generated`` macro.
    This adds some information to the AST that lets the compiler know that
    this is a generated function.
 
 2. In the body of the generated function you only have access to the
-   *types* of the arguments, not their values.
+   *types* of the arguments -- not their values -- and any function that
+   was defined *before* the definition of generated function.
```
Comment (mauro3) [inline doc/manual/metaprogramming.rst]: "was defined *before* the definition of the generated function."

File: doc/manual/metaprogramming.rst [hunk 9]
```diff
@@ -1028,12 +1037,40 @@ copy them*, for the following reasons:
 * the ``foo`` function has side-effects, and it is undefined exactly when,
```
Comment (Sacha0) [inline doc/manual/metaprogramming.rst]: As `@mauro3` suggested in https://github.com/JuliaLang/julia/pull/18679/files#r80905058, perhaps also mention the issues immediately alongside the corresponding examples?

File: doc/manual/metaprogramming.rst [hunk 10]
```diff
@@ -1028,12 +1037,40 @@ copy them*, for the following reasons:
...
+  4. Taking any locks: C code you call out to may use locks internally,
+     (for example, it is not problematic to call ``malloc``,
+     even though most implementations require locks internally)
+     but don't attempt to hold or acquire any while executing Julia code.
+
+  5. Calling any function that is defined after the body of the generated function.
+     This condition is relaxed for incrementally-loaded precompiled modules to
+     allow calling any function in the module.
+
+Alright, now that we have a better understanding for how generated functions
```
Comment (Sacha0) [inline doc/manual/metaprogramming.rst]: for -> of ("now that we have a better understanding of how generated functions work").

File: doc/manual/metaprogramming.rst (hunk 6 above)
Comment (vtjnash) [inline doc/manual/metaprogramming.rst]: `println` isn't OK, it's just a convenient function for showing how they work (and along the way demonstrating why it isn't OK)

File: doc/manual/metaprogramming.rst [hunk 11]
```diff
@@ -1034,7 +1034,7 @@ These examples are hopefully helpful to illustrate how generated functions
 work, both in the definition end and at the call site; however, *don't
 copy them*, for the following reasons:
 
-* the ``foo`` function has side-effects, and it is undefined exactly when,
+* the ``foo`` function has side-effects (the call to `Core.println`), and it is undefined exactly when,
```
Comment (tkelman) [inline doc/manual/metaprogramming.rst]: rst, so double backticks

**Comment by vtjnash** [comment]
rebased to squash (previous head was 8f4d569) and added 94be098bbedb86eb575b17067fa5ded173eff3a8 to fix a pre-existing bug, for backporting (not all of these commits need to be backported, but can review that later)

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs=:master)`

**Comment by jrevels** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18679#issuecomment-251160136) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/ac007972fd8a729bb107c27c47912e083e6de484/1560b17_vs_6c27e24/report.md). cc @jrevels

**Comment by vtjnash** [comment]
nanosoldier says, "merge"?

File: src/gf.c [hunk 12]
```diff
@@ -829,7 +829,7 @@ static jl_method_instance_t *jl_mt_assoc_by_type(jl_methtable_t *mt, jl_datatype
     sig = join_tsig(tt, entry->sig);
     jl_method_instance_t *nf;
     if (!cache) {
-        nf = jl_get_specialized(m, sig, env); // TODO: should be jl_specializations_get_linfo
+        nf = jl_specializations_get_linfo(m, sig, env);
```
Comment (maleadt) [inline src/gf.c]: Just FYI, this change doesn't fix the ASAN issue I mentioned.

**Comment by KristofferC** [comment]
I'm excited for the new worlds to come.

**Pull request merged**